data-plane/receiver-loom/src/test/java/dev/knative/eventing/kafka/broker/receiverloom/LoomKafkaProducerTest.java

```diff
@@ -0,0 +1,212 @@
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(VertxExtension.class)
+public class LoomKafkaProducerTest {
+
+    private Vertx vertx;
+    private LoomKafkaProducer<String, Integer> producer;
+
+    @BeforeEach
+    public void setUp() {
+        vertx = Vertx.vertx();
+        producer = new LoomKafkaProducer<>(vertx, mockProducer());
+    }
+
+    @AfterEach
+    public void tearDown() {
+        producer.close();
+    }
+
+    @Test
+    public void testConcurrentRecordSendCountAndSequence() throws InterruptedException {
+
+        // Set up the test parameters
+        int numRecords = 100000;
+        AtomicInteger counter = new AtomicInteger();
+        CountDownLatch latch = new CountDownLatch(numRecords);
+        List<Integer> receivedOrder = new ArrayList<>();
+
+        // Send the records
+        for (int i = 0; i < numRecords; i++) {
+            int finalI = i;
+            sendRecord(counter, latch, receivedOrder, finalI);
+        }
+
+        latch.await();
+
+        // Verify the number of records sent
+        assertEquals(numRecords, counter.get());
+
+        // Verify the order of received records
+        for (int i = 0; i < numRecords; i++) {
+            assertEquals(i, receivedOrder.get(i));
+        }
+    }
+
+    @Test
+    public void testSendAfterClose(VertxTestContext testContext) {
+        // Close the producer before sending a record
+        producer.close();
+
+        // Attempt to send a record after the producer is closed
+        ProducerRecord<String, Integer> record = new ProducerRecord<>("test", "sequence number", 123);
+        Future<RecordMetadata> sendFuture = producer.send(record);
+
+        // Verify that the sendFuture fails with the expected error message
+        sendFuture.onComplete(ar -> {
+            testContext.verify(() -> {
+                assertTrue(ar.failed());
+                assertEquals("Producer is closed", ar.cause().getMessage());
+                testContext.completeNow();
+            });
+        });
+    }
+
+    @Test
+    public void testCloseIsWaitingForEmptyQueue(VertxTestContext testContext) {
+        // Send multiple records
+        int numRecords = 100000;
+        List<Future<RecordMetadata>> sendFutures = new ArrayList<>();
+        for (int i = 0; i < numRecords; i++) {
+            ProducerRecord<String, Integer> record = new ProducerRecord<>("test", "sequence number", i);
+            sendFutures.add(producer.send(record));
+        }
+
+        // Close the producer and wait for the event queue to be empty
+        producer.close().onComplete(ar -> {
```

Comment:
We need to set what happens on failure with `.onFailure(testContext::failNow)` so that we get a more descriptive error message